npm/ipsm/ipsm.go

```diff
@@ -0,0 +1,439 @@
+package ipsm
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+
+	"github.com/Azure/azure-container-networking/log"
+	"github.com/Azure/azure-container-networking/npm/util"
+)
+
+type ipsEntry struct {
+	operationFlag string
+	name          string
+	set           string
+	spec          string
+}
+
+// IpsetManager stores ipset states.
+type IpsetManager struct {
+	listMap map[string]*Ipset //tracks all set lists.
+	setMap  map[string]*Ipset //label -> []ip
+}
+
+// Ipset represents one ipset entry.
+type Ipset struct {
+	name       string
+	elements   []string
+	referCount int
+}
+
+// NewIpset creates a new instance for Ipset object.
+func NewIpset(setName string) *Ipset {
+	return &Ipset{
+		name: setName,
+	}
+}
+
+// NewIpsetManager creates a new instance for IpsetManager object.
+func NewIpsetManager() *IpsetManager {
+	return &IpsetManager{
+		listMap: make(map[string]*Ipset),
+		setMap:  make(map[string]*Ipset),
+	}
+}
+
+// Exists checks if an element exists in setMap/listMap.
+func (ipsMgr *IpsetManager) Exists(key string, val string, kind string) bool {
+	m := ipsMgr.setMap
+	if kind == util.IpsetSetListFlag {
+		m = ipsMgr.listMap
+	}
+
+	if _, exists := m[key]; !exists {
+		return false
+	}
+
+	for _, elem := range m[key].elements {
+		if elem == val {
+			return true
+		}
+	}
+
+	return false
+}
+
+func isNsSet(setName string) bool {
+	return !strings.Contains(setName, "-") && !strings.Contains(setName, ":")
+}
+
+// IncrementReferCount increases referCount of a specific ipset by one.
+func (ipsMgr *IpsetManager) IncrementReferCount(setName string) error {
+	if _, exists := ipsMgr.setMap[setName]; !exists {
+		log.Printf("set %s doesn't exist, can't increment", setName)
+		return fmt.Errorf("set %s doesn't exist, can't increment", setName)
+	}
+
+	ipsMgr.setMap[setName].referCount++
+
+	return nil
+}
+
+// DecrementReferCount decreases referCount of a specific ipset by one.
+func (ipsMgr *IpsetManager) DecrementReferCount(setName string) error {
+	if _, exists := ipsMgr.setMap[setName]; !exists {
+		log.Printf("set %s doesn't exist, can't increment", setName)
+		return fmt.Errorf("set %s doesn't exist, can't increment", setName)
+	}
+
+	ipsMgr.setMap[setName].referCount--
+
+	return nil
+}
+
+// NotReferredByNwPolicy checks if a specific ipset is referred by any network policy.
+func (ipsMgr *IpsetManager) NotReferredByNwPolicy(setName string) bool {
+	return ipsMgr.setMap[setName].referCount == 0
```

Comment:
what if setName is not present in the setMap?

Reply:
unused func. It is removed.